src/framework/components/element/pointer-events.js

```diff
@@ -0,0 +1,190 @@
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventUp", [ point ]);
+            this.fire(pc.POINTEREVENT_CLICK, point);
+            this.fire(pc.POINTEREVENT_UP, point);
+        },
+
+        // Fires pointer leave event and also makes all children do so.
+        _ensurePointerLeaveEvent: function(point) {
+            this._pointerOver = false;
+            this.fire(pc.POINTEREVENT_LEAVE, point); 
+
+            this._passPointerEventToChildren( "_ensurePointerLeaveEvent", [ point ]);
+        },
+
+        // Handles "move" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventMove: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                if (this._pointerOver) {
+                    this._ensurePointerLeaveEvent(point);            
+                }
+
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventMove", [ point ]);
+
+            if (!this._pointerOver) {
+                this._pointerOver = true;
+                this.fire(pc.POINTEREVENT_ENTER, point);                
+            }
+
+            this.fire(pc.POINTEREVENT_MOVE, point);
+        },
+
+        // Handles "scroll" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventScroll: function(point, amount) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventScroll", [ point, amount ]);
+            this.fire(pc.POINTEREVENT_SCROLL, point, amount);
+        },
+
+        // Mouse-specific event handler.
+        _onMouseDown: function(mouseEvent) {
+            this._pointerEventDown( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
+        },
+
+        // Mouse-specific event handler.
+        _onMouseUp: function(mouseEvent) {
+            this._pointerEventUp( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
+        },
+
+        // Mouse-specific event handler.
+        _onMouseMove: function(mouseEvent) {
+            this._pointerEventMove( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
+        },
+
+        // Mouse-specific event handler.
+        _onMouseMove: function(mouseEvent) {
+            this._pointerEventMove( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
+        },
+
+        // Touch-specific event handler.
+        _onMouseWheel: function(mouseEvent) {
+            this._pointerEventScroll( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ), mouseEvent.wheel );
+        },
+
+        // Touch-specific event handler.
+        _onTouchUp: function(touchEvent) {
+            var touch = touchEvent.changedTouches[0];
```

Comment:
What about multi-touch? And it could have been not the first finger that clicked on button.